<a href="https://colab.research.google.com/github/sdasmaha/Ml_Engineers_445/blob/master/Diabetes_Notebook_Exercise1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 Notebook Excercise #1 Part2 
 SRINJOY DASMAHAPATRA

 IDE => Google Colab


#Importing Libraries

In [681]:
import numpy as np
import pandas as pd

#Mounting the Google Drive


In [682]:
from google.colab import drive
drive.mount("/content/gdrive")

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).



#Loading the Dataset


In [683]:
diabetes_df = pd.read_csv('/content/gdrive/My Drive/Colab Notebooks/BAJWA-ML/diabetes_dataset.csv')

# *2.1*

**A**

In [684]:
print(diabetes_df.shape) 

(768, 9)


In [685]:
print(diabetes_df.axes)

[RangeIndex(start=0, stop=768, step=1), Index(['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin',
       'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'],
      dtype='object')]


In [686]:
print(diabetes_df.dtypes)

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object


**B**

In [687]:
diabetes_df.head(10)

Pregnancies  Glucose  BloodPressure  ...  DiabetesPedigreeFunction  Age  Outcome
0            6      148             72  ...                     0.627   50        1
1            1       85             66  ...                     0.351   31        0
2            8      183             64  ...                     0.672   32        1
3            1       89             66  ...                     0.167   21        0
4            0      137             40  ...                     2.288   33        1
5            5      116             74  ...                     0.201   30        0
6            3       78             50  ...                     0.248   26        1
7           10      115              0  ...                     0.134   29        0
8            2      197             70  ...                     0.158   53        1
9            8      125             96  ...                     0.232   54        1

[10 rows x 9 columns]

**C**    Based on the knowledge of the dataset and its stated usage, we are dealing with supervised problem. 

Since in a supervised learning model, the algorithm learns on a labeled dataset, providing an answer key that the algorithm can use to evaluate its accuaracy on training data. 

On the contrary the unsupervised model works on unlabeled data that the algorithm tries to make sense of by extracting features and patterns on its own.


**D** The independent variables in this case are  labels- 

1. Pregnancies	
2. Glucose
3. BloodPressure
4. SkinThickness
5. Insulin
6. BMI
7. DiabetesPedigreeFunction
8. Age




This is so because these variable/labels embedded in the algorithm 
increase or decrease the probability/likhehood of the event
"*Outcome*".


**E** The dependent variable in this case is the label *Outcome* since it is the conclusive/final event or variable that predicts the likelihood of the occurence of diabetes.

**F** Here the categorical variables are - *

1.  Outcome



This is so because a categorical variable is a variable that can take on one of a limited, and usually fixed, number of possible values, assigning each individual or other unit of observation to a particular group or nominal category on the basis of some qualitative property.



**G** The categorical variables follow *Label Encoding* where each label is mapped to a unique quantitative value.



**H**

In [688]:
pd.cut(x=diabetes_df['Age'], bins=[21, 40, 60, 110]).value_counts()

(21, 40]     511
(40, 60]     167
(60, 110]     27
Name: Age, dtype: int64

Young Adults : 511 samples

1.   Young Adults : 511 samples
2.   Middle-ages Adults : 167 samples
3.   Old-aged Adults: 27



In [689]:
diabetes_df.dtypes

Pregnancies                   int64
Glucose                       int64
BloodPressure                 int64
SkinThickness                 int64
Insulin                       int64
BMI                         float64
DiabetesPedigreeFunction    float64
Age                           int64
Outcome                       int64
dtype: object

# *2.2*

In [690]:
diabetes_df.isnull().sum()

Pregnancies                 0
Glucose                     0
BloodPressure               0
SkinThickness               0
Insulin                     0
BMI                         0
DiabetesPedigreeFunction    0
Age                         0
Outcome                     0
dtype: int64

We can clearly see that we have no missing values.

In [691]:
diabetes_df['Pregnancies'].where(diabetes_df['Pregnancies'] == 0).value_counts()

0.0    111
Name: Pregnancies, dtype: int64

In [692]:
diabetes_df['Glucose'].where(diabetes_df['Glucose'] == 0).value_counts()


0.0    5
Name: Glucose, dtype: int64

In [693]:
diabetes_df['BloodPressure'].where(diabetes_df['BloodPressure'] == 0).value_counts()


0.0    35
Name: BloodPressure, dtype: int64

In [694]:
diabetes_df['SkinThickness'].where(diabetes_df['SkinThickness'] == 0).value_counts()


0.0    227
Name: SkinThickness, dtype: int64

In [695]:
diabetes_df['Insulin'].where(diabetes_df['Insulin'] == 0).value_counts()

0.0    374
Name: Insulin, dtype: int64

In [696]:
diabetes_df['BMI'].where(diabetes_df['BMI'] == 0).value_counts()


0.0    11
Name: BMI, dtype: int64

In [698]:
diabetes_df['DiabetesPedigreeFunction'].where(diabetes_df['DiabetesPedigreeFunction'] == 0).value_counts()


Series([], Name: DiabetesPedigreeFunction, dtype: int64)

In [699]:
diabetes_df['Age'].where(diabetes_df['Age'] < 0).value_counts()


Series([], Name: Age, dtype: int64)

We can Clearly See we have so many values that correspond to numeric 0 which makes the dataset inconsistent.

Now converting the inconsistent(O numeric)values with NaN.

In [700]:
diabetes_df['Glucose'].replace(0, np.NAN, inplace = True)
diabetes_df['BloodPressure'].replace(0, np.NAN, inplace=True)
diabetes_df['SkinThickness'].replace(0, np.NAN, inplace=True)
diabetes_df['Insulin'].replace(0, np.NAN, inplace=True)
diabetes_df['BMI'].replace(0, np.NAN, inplace=True)
diabetes_df['Age'].replace(0, np.NAN, inplace=True)
diabetes_df.head(20)


Pregnancies  Glucose  BloodPressure  ...  DiabetesPedigreeFunction  Age  Outcome
0             6    148.0           72.0  ...                     0.627   50        1
1             1     85.0           66.0  ...                     0.351   31        0
2             8    183.0           64.0  ...                     0.672   32        1
3             1     89.0           66.0  ...                     0.167   21        0
4             0    137.0           40.0  ...                     2.288   33        1
5             5    116.0           74.0  ...                     0.201   30        0
6             3     78.0           50.0  ...                     0.248   26        1
7            10    115.0            NaN  ...                     0.134   29        0
8             2    197.0           70.0  ...                     0.158   53        1
9             8    125.0           96.0  ...                     0.232   54        1
10            4    110.0           92.0  ...                     0.191   30        0
11           10    168.0           74.0  ...                     0.537   34        1
12           10    139.0           80.0  ...                     1.441   57        0
13            1    189.0           60.0  ...                     0.398   59        1
14            5    166.0           72.0  ...                     0.587   51        1
15            7    100.0            NaN  ...                     0.484   32        1
16            0    118.0           84.0  ...                     0.551   31        1
17            7    107.0           74.0  ...                     0.254   31        1
18            1    103.0           30.0  ...                     0.183   33        0
19            1    115.0           70.0  ...                     0.529   32        1

[20 rows x 9 columns]

The dataset didn't have missing values in it. We can clearly see here the inconsistent values were a bunch of zeroes by itself. We replaced them with NaN.

# *2.3*

The Non Categorical columns in this dataset are -


1.   Pregnancies
2.   Glucose
3.   BloodPressure
4.   SkinThickness
5.   Insulin
6.   BMI
7.   DiabetesPedigreeFunction
8.   Age




In [671]:
diabetes_df = diabetes_df[['Pregnancies','Glucose','BloodPressure','SkinThickness','Insulin','BMI','DiabetesPedigreeFunction','Age']]

In [672]:
avg = diabetes_df.mean()

In [673]:
stdv = diabetes_df.std()


In [674]:
standardized = (diabetes_df - avg) / stdv

In [675]:
standardized.head(20)

Pregnancies   Glucose  ...  DiabetesPedigreeFunction       Age
0      0.639530  0.847771  ...                  0.468187  1.425067
1     -0.844335 -1.122665  ...                 -0.364823 -0.190548
2      1.233077  1.942458  ...                  0.604004 -0.105515
3     -0.844335 -0.997558  ...                 -0.920163 -1.040871
4     -1.141108  0.503727  ...                  5.481337 -0.020483
5      0.342757 -0.153085  ...                 -0.817546 -0.275580
6     -0.250789 -1.341602  ...                 -0.675693 -0.615709
7      1.826623 -0.184362  ...                 -1.019762 -0.360612
8     -0.547562  2.380333  ...                 -0.947326  1.680164
9      1.233077  0.128406  ...                 -0.723983  1.765196
10     0.045984 -0.340746  ...                 -0.847727 -0.275580
11     1.826623  1.473307  ...                  0.196553  0.064549
12     1.826623  0.566280  ...                  2.924962  2.020293
13    -0.844335  2.130119  ...                 -0.222970  2.190358
14     0.342757  1.410753  ...                  0.347461  1.510099
15     0.936304 -0.653513  ...                  0.036591 -0.105515
16    -1.141108 -0.090532  ...                  0.238807 -0.190548
17     0.936304 -0.434576  ...                 -0.657584 -0.190548
18    -0.844335 -0.559683  ...                 -0.871873 -0.020483
19    -0.844335 -0.184362  ...                  0.172408 -0.105515

[20 rows x 8 columns]

In [676]:
standardized.mean()

Pregnancies                 2.977942e-17
Glucose                    -4.192248e-18
BloodPressure              -7.083454e-18
SkinThickness               3.050222e-17
Insulin                     1.850372e-17
BMI                         2.966016e-15
DiabetesPedigreeFunction    1.913039e-15
Age                         2.198762e-16
dtype: float64

In [677]:
standardized.var()

Pregnancies                 1.0
Glucose                     1.0
BloodPressure               1.0
SkinThickness               1.0
Insulin                     1.0
BMI                         1.0
DiabetesPedigreeFunction    1.0
Age                         1.0
dtype: float64

We can clearly see the mean is ~ 0 and the variance is 1 for all the non categorical columns.

# *2.4*

Changing the *Outcome* is encoded using one hot encoding.

In [701]:
diabetes_df = diabetes_df.rename(columns = {"Outcome": "1"})

In [702]:
diabetes_df["0"] = diabetes_df["1"].map(lambda x: 0 if x == 1 else (1))

In [704]:
diabetes_df.head(20)

Pregnancies  Glucose  BloodPressure  ...  Age  1  0
0             6    148.0           72.0  ...   50  1  0
1             1     85.0           66.0  ...   31  0  1
2             8    183.0           64.0  ...   32  1  0
3             1     89.0           66.0  ...   21  0  1
4             0    137.0           40.0  ...   33  1  0
5             5    116.0           74.0  ...   30  0  1
6             3     78.0           50.0  ...   26  1  0
7            10    115.0            NaN  ...   29  0  1
8             2    197.0           70.0  ...   53  1  0
9             8    125.0           96.0  ...   54  1  0
10            4    110.0           92.0  ...   30  0  1
11           10    168.0           74.0  ...   34  1  0
12           10    139.0           80.0  ...   57  0  1
13            1    189.0           60.0  ...   59  1  0
14            5    166.0           72.0  ...   51  1  0
15            7    100.0            NaN  ...   32  1  0
16            0    118.0           84.0  ...   31  1  0
17            7    107.0           74.0  ...   31  1  0
18            1    103.0           30.0  ...   33  0  1
19            1    115.0           70.0  ...   32  1  0

[20 rows x 10 columns]

# *2.5*

In [705]:
Glucose_median=diabetes_df['Glucose'].median()
diabetes_df['Glucose'].replace(np.NAN, Glucose_median, inplace= True)

BloodPressure_median=diabetes_df['BloodPressure'].median()
diabetes_df['BloodPressure'].replace(np.NAN, BloodPressure_median, inplace= True)

SkinThickness_median=diabetes_df['SkinThickness'].median()
diabetes_df['SkinThickness'].replace(np.NAN, SkinThickness_median, inplace= True)

Insulin_median=diabetes_df['Insulin'].median()
diabetes_df['Insulin'].replace(np.NAN,Insulin_median, inplace= True)

BMI_median=diabetes_df['BMI'].median()
diabetes_df['BMI'].replace(np.NAN,BMI_median, inplace= True)

Age_median=diabetes_df['Age'].median()
diabetes_df['Age'].replace(np.NAN,Age_median, inplace= True)

diabetes_df.head(20)


Pregnancies  Glucose  BloodPressure  ...  Age  1  0
0             6    148.0           72.0  ...   50  1  0
1             1     85.0           66.0  ...   31  0  1
2             8    183.0           64.0  ...   32  1  0
3             1     89.0           66.0  ...   21  0  1
4             0    137.0           40.0  ...   33  1  0
5             5    116.0           74.0  ...   30  0  1
6             3     78.0           50.0  ...   26  1  0
7            10    115.0           72.0  ...   29  0  1
8             2    197.0           70.0  ...   53  1  0
9             8    125.0           96.0  ...   54  1  0
10            4    110.0           92.0  ...   30  0  1
11           10    168.0           74.0  ...   34  1  0
12           10    139.0           80.0  ...   57  0  1
13            1    189.0           60.0  ...   59  1  0
14            5    166.0           72.0  ...   51  1  0
15            7    100.0           72.0  ...   32  1  0
16            0    118.0           84.0  ...   31  1  0
17            7    107.0           74.0  ...   31  1  0
18            1    103.0           30.0  ...   33  0  1
19            1    115.0           70.0  ...   32  1  0

[20 rows x 10 columns]

In [706]:
from google.colab import files
diabetes_df.to_csv('diabetes_dataset_processed.csv') 
files.download('diabetes_dataset_processed.csv')

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>